app/plugin.go

```diff
@@ -182,6 +190,78 @@ func (a *App) InitPlugins(pluginDir, webappPluginDir string) {
 	a.SyncPluginsActiveState()
 }
 
+// SynchPlugins will synchronize the plugins installed locally
+// with the plugin bundles available in the file store.
+func (a *App) SynchPlugins() *model.AppError {
+	mlog.Info("Synching plugins with the file store ...")
+
+	pluginDirs, err := ioutil.ReadDir(*a.Config().PluginSettings.Directory)
+	if err != nil {
+		return model.NewAppError("SynchPlugins", "app.plugin.synch.read_local_folder.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	// Remove plugins locally to synch with file store.
+	for _, dir := range pluginDirs {
+		if !dir.IsDir() {
+			continue
+		}
+		dirFullPath := filepath.Join(*a.Config().PluginSettings.Directory, dir.Name())
+		files, readErr := ioutil.ReadDir(dirFullPath)
+
+		if readErr != nil {
+			mlog.Error(fmt.Sprintf("Error reading local plugin directoy: %v. Skipped for synch.", dir.Name()), mlog.Err(readErr))
+			continue
+		}
+		// Only handle managed plugins with .filestore flag file.
+		managed := false
+		for _, f := range files {
+			if f.Name() == ".filestore" {
+				managed = true
+
+				mlog.Debug(fmt.Sprintf("Plugin Synch: Uninstalling %v plugin locally", dir.Name()))
+				if err := a.removePlugin(dir.Name(), true); err != nil {
+					mlog.Error(fmt.Sprintf("Plugin Synch: Error uninstalling managed plugin: %v.", dir.Name()), mlog.Err(err))
+				}
+				break
+			}
+		}
+
+		if !managed {
+			mlog.Warn(fmt.Sprintf("Found unmanaged plugin: %v. Ignoring in plugin synch with the filestore.", dir.Name()))
+		}
+	}
+
+	// Install plugins from the file store.
+	exists, existsErr := a.FileExists("./plugins")
+
+	if existsErr != nil {
+		return model.NewAppError("SynchPlugins", "app.plugin.synch.check_filestore.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	if exists {
+		fileStorepaths, listDirErr := a.ListAll("./plugins")
+		if listDirErr != nil {
+			return model.NewAppError("SynchPlugins", "app.plugin.synch.list_filestore.app_error", nil, listDirErr.Error(), http.StatusInternalServerError)
+		}
+
+		for _, path := range fileStorepaths {
+			if strings.HasSuffix(path, ".tar.gz") {
+				fileBytes, fileReaderErr := a.ReadFile(filepath.Join("./", path))
```

Comment:
Unnecessary copy here. Can we use the reader interface and just pass the reader to InstallPlugin?

Reply:
@crspeller  Went with `ReadFile` because the Reader consumes the file that is stored in the file store and leaves it empty, even after Seek(0,0) 🤔 Any suggestions on this?

Reply:
We only need to read it once, though, right?

Reply:
Yes, it's only read once per bundle at startup time. This is not executed during regular installation.

Reply:
Hmm, then I'm confused what would leave the reader empty in that case. Calling `app.FileReader` should return a `filestore.ReadCloseSeeker` ready for use: is the problem that you're seeing it behave otherwise?

Reply:
The problem that I was having is that is was running into a conflict with the stream and the source when it tried to install again to the filestore. I tested it again and now it's not an issue with the change to do just `installPluginLocally`.  👍
It is back now to a Reader. Thanks!